klippa/tests/integration_test.rs

```diff
@@ -0,0 +1,538 @@
+                    "Yes" => {
+                        self.case.iup_optimize.push(true);
+                    }
+                    "No" => {
+                        self.case.iup_optimize.push(false);
+                    }
+                    _ => {
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl SubsetTestCase {
+    pub fn new(path: &Path) -> Self {
+        let parser = TestCaseParser::new();
+        parser.parse(path)
+    }
+
+    pub fn run(&self) {
+        let output_temp_dir = TempDir::new_in(".", "klippa_test").unwrap();
+        let output_dir = output_temp_dir.path();
+        for font in &self.fonts {
+            //TODO: add support for profiles later
+            //for profile in &self.profiles {
+            //}
+            for subset in &self.subsets {
+                //TODO: add support for instances/iup_options
+                self.run_one_test(font, subset, output_dir);
+            }
+        }
+    }
+
+    pub fn gen_expected_output(&self) {
+        //let output_dir = PathBuf::from("./qxliu_test");
+        //fs::create_dir("./qxliu_test").unwrap();
+        let output_temp_dir = TempDir::new_in(".", "klippa_test").unwrap();
+        let output_dir = output_temp_dir.path();
+        for font in &self.fonts {
+            //TODO: add support for profiles later
+            //for profile in &self.profiles {
+            //}
+            for subset in &self.subsets {
+                //TODO: add support for instances/iup_options
+                self.gen_expected_output_for_one_test(font, subset, output_dir);
+            }
+        }
+        let expected_dir = Path::new(TEST_DATA_DIR)
+            .join("expected")
+            .join(&self.expected_dir);
+        fs::rename(output_dir, expected_dir).unwrap();
+    }
+
+    fn run_one_test(&self, font: &str, subset: &str, output_dir: &Path) {
+        //TODO: re-enable subset="*" once populate_unicodes_to_retain supports *
+        if subset == "*" {
+            return;
+        }
+        let subset_font_name = gen_subset_font_name(font, subset);
+        let output_file = output_dir.join(&subset_font_name);
+        gen_subset_font_file(font, subset, &output_file);
+
+        let expected_file = Path::new(TEST_DATA_DIR)
+            .join("expected")
+            .join(&self.expected_dir)
+            .join(&subset_font_name);
+        compare_with_expected(output_dir, &output_file, &expected_file);
+    }
+
+    fn gen_expected_output_for_one_test(&self, font: &str, subset: &str, output_dir: &Path) {
+        //TODO: re-enable subset="*" once populate_unicodes_to_retain supports *
+        if subset == "*" {
+            return;
+        }
+
+        let subset_font_name = gen_subset_font_name(font, subset);
+        let output_file = output_dir.join(&subset_font_name);
+        gen_subset_font_file(font, subset, &output_file);
+
+        assert_has_ttx_exec();
```

Comment:
I might just move this into the top of the main test body, where we just run it once before we do anything else?

Reply:
we cannot, because we generate multiple subset font files in a nested loop for one test